velox/functions/sparksql/specialforms/DecimalRound.cpp

```diff
@@ -0,0 +1,208 @@
+      const TypePtr& resultType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const {
+    context.ensureWritable(rows, resultType, result);
+    result->clearNulls(rows);
+    return result->asUnchecked<FlatVector<R>>()->mutableRawValues();
+  }
+
+  template <typename R /* Result */, typename A /* Argument */>
+  void applyRoundRows(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      uint8_t inputPrecision,
+      uint8_t inputScale,
+      uint8_t resPrecision,
+      uint8_t resScale,
+      int32_t scale,
+      const TypePtr& resultType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const {
+    VELOX_USER_CHECK(
+        args[0]->isConstantEncoding() || args[0]->isFlatEncoding(),
+        "Single-arg deterministic functions receive their only argument as flat or constant vector.");
+    auto rawResults = prepareResults<R>(rows, resultType, context, result);
+    if (args[0]->isConstantEncoding()) {
+      // Fast path for constant vectors.
+      auto constant = args[0]->asUnchecked<ConstantVector<A>>()->valueAt(0);
+      context.applyToSelectedNoThrow(rows, [&](auto row) {
+        applyRound<R, A>(
+            rawResults[row],
+            constant,
+            inputPrecision,
+            inputScale,
+            resPrecision,
+            resScale,
+            scale);
+      });
+    } else {
+      // Fast path for flat.
+      auto flatA = args[0]->asUnchecked<FlatVector<A>>();
+      auto rawA = flatA->mutableRawValues();
+      context.applyToSelectedNoThrow(rows, [&](auto row) {
+        applyRound<R, A>(
+            rawResults[row],
+            rawA[row],
+            inputPrecision,
+            inputScale,
+            resPrecision,
+            resScale,
+            scale);
+      });
+    }
+  }
+
+  template <typename R, typename A>
+  inline void applyRound(
+      R& r,
+      const A& a,
+      uint8_t inputPrecision,
+      uint8_t inputScale,
+      uint8_t resPrecision,
+      uint8_t resScale,
+      int32_t scale) const {
+    if (scale >= 0) {
+      auto rescaledValue = DecimalUtil::rescaleWithRoundUp<A, R>(
+          a, inputPrecision, inputScale, resPrecision, resScale);
+      VELOX_DCHECK(rescaledValue.has_value());
+      r = rescaledValue.value();
+    } else {
+      auto reScaleFactor = DecimalUtil::kPowersOfTen[inputScale - scale];
+      DecimalUtil::divideWithRoundUp<R, A, int128_t>(
+          r, a, reScaleFactor, false, 0, 0);
+      r = r * DecimalUtil::kPowersOfTen[-scale];
+    }
+  }
+};
+}; // namespace
+
+TypePtr DecimalRoundCallToSpecialForm::resolveType(
+    const std::vector<TypePtr>& argTypes) {
+  VELOX_FAIL("Decimal round function does not support type resolution.");
+}
+
+exec::ExprPtr DecimalRoundCallToSpecialForm::constructSpecialForm(
+    const TypePtr& type,
+    std::vector<exec::ExprPtr>&& compiledChildren,
+    bool trackCpuUsage,
+    const core::QueryConfig& config) {
+  VELOX_USER_CHECK(
```

Comment:
split this check into 2 for better error messages

VELOX_USER_CHECK_GT(compiledChildren.size(), 0, "decimal_round expects one or two arguments")
VELOX_USER_CHECK_LE(compiledChildren.size(), 2, "decimal_round expects one or two arguments")